Misc fixes and a couple tests

TylerLeonhardt opened the pull request:
This fixes a couple issues I've found on macOS:

1. Setting and then retrieving a secret didn't work until after a few ms... apparently this was because secrets that are not "saved to lastpass" yet have a date time in the output of `lpass ls`. I modified the regex to include this.
2. Even though we were routing errors in the error stream to `/dev/null` I still would see the text `Error: Can't find the account` (or something like that). To fix that, I routed all streams to output and handled errors accordingly
3. `$BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($Secret.$k)` etc... didn't work on macOS... like it would only return the first character of the secret... very odd... to fix this I use `[System.Net.NetworkCredential]::new("", $Secret.NoteType).Password`

This also adds a couple tests I stole from Steve's KeyChain vault... some are disabled for now. But we can bring those back in time...

@itfranck interested in reviewing?

TylerLeonhardt committed:
misc fixes and a couple tests

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -57,14 +58,14 @@ function Invoke-lpass {
     if ($lpassCommand -ne '' -and ((& "$lpassCommand" $lpassPath --version ) -like 'LastPass CLI*') ) {
         if ($InputObject) {
-            return $InputObject | & "$lpassCommand" $lpassPath @Arguments 
+            return $InputObject | & "$lpassCommand" $lpassPath @Arguments *>&1
```
@TylerLeonhardt  By redirecting everything, it also kill that prompt. 

![image](https://user-images.githubusercontent.com/1980296/98455026-5eaa3e80-2139-11eb-9792-ab68e080f3ca.png)

More than that, there's no message to indicate that I need to login and instead it return an empty secret

![image](https://user-images.githubusercontent.com/1980296/98455044-b052c900-2139-11eb-8c1f-4fccf9a0b96b.png)

Valid or invalid secrets, everything is now an empty string. 

I encountered this initially when working on the previous PR and that's why I did settle down on doing it only on the `lpass show` within Set-Secret rather than everywhere.

TylerLeonhardt replied:
Hmmm what if we just redirect the error stream instead of _all_ streams?

itfranck replied:
@TylerLeonhardt This seems to come from the error stream. `2>&1` do not produce the prompt either. 😞

itfranck replied:
@TylerLeonhardt Not sure what happened with "past me" but I just tested again  and `2>&1` did work.. ¯\_(ツ)_/¯

itfranck replied:
I submitted another PR to this branch with the suggested changes.

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -174,25 +184,24 @@ function Set-Secret
     try {
-        $res = Invoke-lpass 'show', '--sync=now', '--name', $Name, '2>/dev/null' -ErrorAction Stop 
-        $SecretExists = $null -ne $res 
+        $res = Invoke-lpass 'show', '--sync=now', '--name', $Name
```
Related to the other comment. If you were to remove the redirect from invoke-lpass, you'd need to get it back here.

Except... I did try using : 

```
        $res = Invoke-lpass 'show', '--sync=now', '--name', $Name, '*>&1'
```

but it does show the error message. 
Any idea how to pass it down properly ?
Otherwise I'm thinking it might need to be cooked in Invoke-lpass and made available through a switch so it can be used on a in-demand basis (aka here)

TylerLeonhardt replied:
Yeah that's what I was trying to avoid as it makes Invoke-lpass more complicated but that might be what it takes...

itfranck commented:
@TylerLeonhardt See my comments. 

Using redirect stream everywhere also disable the login prompt and empty secret get returned. 
I had this issue too while doing the original PR and this is why I ended up adding the redirect stream only for the Show call within Set-Secret.

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -174,25 +184,24 @@ function Set-Secret
-        $SecretExists = $null -ne $res 
+        $res = Invoke-lpass 'show', '--sync=now', '--name', $Name
+        # We use ToString() here to turn the ErrorRecord into a string if we got an ErrorRecord
+        $SecretExists = $res.ToString() -ne "Error: Could not find specified account(s)."
```
I think I would add a null check just in case. 
What do you think ?
```suggestion
        $SecretExists = $null -ne $res -and $res.ToString() -ne "Error: Could not find specified account(s)."
```

TylerLeonhardt committed:
add null check

Co-authored-by: Francis Mercier <Francis@nzt.ca>

itfranck committed:
Bugfixes/verboseflag + Notes edge case (proper) (#13)

* Update SecretManagement.LastPass.Extension.psm1

* formatting

* Verbose handling with etter syntax ;)

* Update SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1

Co-authored-by: Tyler James Leonhardt <tylerl0706@gmail.com>

* Update SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1

Co-authored-by: Tyler James Leonhardt <tylerl0706@gmail.com>

Co-authored-by: Tyler James Leonhardt <tylerl0706@gmail.com>

itfranck committed:
Buffix/ignore errors elegantly (#14)

* Comment is obsolete.

* Ignore errors

* Missing silently continue.

* Define lpass messages as constant + Set-Secret add. validation

* formatting

* formatting

* nit

* Switched order because ABC

* fix

* formatting

TylerLeonhardt commented:
my tests aren't looking good after that last change:

```
❯❯❯ ./build.ps1 -Test


Starting discovery in 1 files.
Discovery finished in 16ms.
Error: Could not find specified account(s).
WARNING: Querying the secret tests/8c148c24-711e-4795-abba-902ccd07261a produced an unexpected result of $Null
Error: Could not find specified account(s).
Error: Could not find specified account(s).
WARNING: Querying the secret tests/f52be38a-3917-4429-8840-b3314a6eebec produced an unexpected result of $Null
Error: Could not find specified account(s).
Error: Could not find specified account(s).
WARNING: Querying the secret tests/15798b20-1cfc-4187-a34a-e7abba4a6d85 produced an unexpected result of $Null
Error: Could not find specified account(s).
Error: Could not find specified account(s).
WARNING: Querying the secret tests/692e4c58-d005-4e73-95bb-6c28b584c773 produced an unexpected result of $Null
```

itfranck commented:
@TylerLeonhardt I don't have this issue (Windows  / WSL)  but it look clear that the error suppression do not work as expected on your environment (Mac? Linux?) 

The error is never redirected `2>&1`  and therefore is displayed.
It causes the warning to be displayed since `$res` is `$null` instead of containing either the secret name or the redirected error. 

I can't reproduce that specific issue on my end though :(
Looks like this is a good starting point : https://github.com/PowerShell/PowerShell/issues/4002

If there's a different way to write the redirect for Mac to make it work, we could then combine that and a OS check and apply the proper redirect.

```
$IsMacOS
$IsLinux
$IsWindows
```

If you can find a proper redirect that work for your OS type, then I can test it on Windows end and we can either implement the globally correct redirect (if any) or on a "per OS"  basis.

TylerLeonhardt commented:
I'll give it a go on my stream today. https://www.twitch.tv/TylerLeonhardt ~30min from now

TylerLeonhardt committed:
changes to fix errors showing up

TylerLeonhardt commented:
@itfranck ok so here's what I changed:

* we now always do `2>&1`
* we throw the wellknown exception (part of SecretManagement contract I guess) called `PasswordRequiredException` when the user is not logged in
* we let SM handle unhandled exceptions (this is done pretty well)
* I added a test for that ^

I wish these tests could be in a CI but that requires more tinkering.

Check the [last commit](https://github.com/TylerLeonhardt/SecretManagement.LastPass/pull/11/commits/476a2991ec380d9519e7e7056ba02aa05724493c) to see those changes.

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -34,11 +37,20 @@ $DefaultNoteTypeMap = @{
     'Wi-Fi Password'    = 'wifi'
 } 
+
+$LoggedOutMessage = 'Error: You don''t apprear to be logged in. Perhaps you need to login with `lpass login`.'
```
@TylerLeonhardt Why remove it from `$lpassMessage` ?
Is it a nit ? 

I thought it was cozy within the other `$lpassMessage`

TylerLeonhardt replied:
I changed the message rather than using the one LastPass had so I pulled it out of the hashtable.

itfranck replied:
@TylerLeonhardt  I kind of liked the old behavior better.
Is this a choice to force the user to login ? 

There's 2 reasons why this message would appear. 
- You are logged out.
- You are logged in but the "active" window closed. 

If you login with lpass --trust email@example.com, you never are really logged out. 

With the old behavior, instead of doing a `lpass status`, the `lpass show` was called directly, which in turn, presented the user with the password prompt. 

So this was kind of a 2 in one compared to now. 
Before: You call `Get-Secret` / `Set-Secret` and the prompt window shows up if your logged in with a disconnected session. 

Currently: You get an error message stating you are not logged in, need to explicitely login again then call the `Get/Set-Secret` cmdlet a second time.

In both scenario, if you are already logged out, then you need to login anyway but 99.9% of the time, I just need to re-establish the session (prompt) and not actually login again. 


You can see the difference in behavior by calling directly
- lpass status
- lpass show

The second one will do the prompt window (if you are logged in but with after 1h of inactivity) 
The first one always require multiple command to be run. 

My suggestion would be to use the show like before (which we use in both get / set) and use the error from that command to determine whether or not to show a "You need to login" message. In the best case, the user will enter his password in the prompt and the cmdlet will continue without having to do a second call to anything else.

itfranck replied:
tl;dr 

When not connected (but logged in) 
**Old flow** 
- User type:  Get-Secret
- User prompted:  Password prompt
- Secret is returned

**Current flow**
- User type: Get-Secret ... 
-  Message: Error: You are currently not logged in ... Try login in with lpass login. 
- User type: lpass login
- User type: Get-Secret ... 
- Secret is returned

Considering the password prompt is to be entered hourly, the second flow become more annoying (You kind of always want to have the login command not far because you'll likely need it). The first (old flow) just prompt you directly for your password and continue afterward. 

In both scenarios, if you are truly logged out, you'll see the error stating you need to login.

TylerLeonhardt replied:
@itfranck I agree with you. Take a look at my last commit

itfranck commented:
I'll commit back to this branch later this evening. 
I found 2 issues in `Get-ComplexSecret` .

I also want to test an edge case regarding NoteType parameter that I might or might not be a problem with the error suppression.

itfranck commented:
> I wish these tests could be in a CI but that requires more tinkering.

Do you know any way to automate the login of the user ?
We'd need to be able to type in automatically the password.

itfranck committed:
Fixes - Complex secrets bug / legacy code (#15)

TylerLeonhardt committed:
go back to previous logic

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -174,51 +202,51 @@ function Set-Secret
+            $NoteTypeArgs += "--note-type=$($Secret.NoteType)"
         }
-       
+        $sb.ToString() | Invoke-lpass 'add', $Name, '--non-interactive', $NoteTypeArgs
```
```suggestion
        $res = $sb.ToString() | Invoke-lpass 'add', $Name, '--non-interactive', $NoteTypeArgs
        if ($null -ne $res)  {
            Write-Error ($res | Out-String)
            return $false
        }
```

Not quite sure that would be the best implementation though but since we suppress the error by default, don't you want to expose the error if it occurs. 

Take for instance this : 

![image](https://user-images.githubusercontent.com/1980296/99024261-30ab6c80-2534-11eb-9614-9cc82d3f6203.png)

If we do nothing, add secret will succeed... 
Even though we received an error from the call. 

To reproduce. 
- Create a secret with a custom type in lastpass (eg: credit card) 
Then 

```
$Boop = Get-Secret -Name "My Credit Card" -AsPlainText -vault lpt -Verbose
$Boop.NoteType = 'sdsd'
Set-Secret -Name 'My Credit Card dupe '  -Secret $Boop -Vault lpt -Verbose
```

TylerLeonhardt replied:
Yep... we still need to let ErrorRecords through that we don't care about. Check out my last commit.

itfranck commented on SecretManagement.LastPass.Extension/SecretManagement.LastPass.Extension.psm1:
```diff
@@ -174,51 +202,51 @@ function Set-Secret
+
+    if ($SecretExists) {
+        Write-Verbose "Editing secret"
+        $sb.ToString() | Invoke-lpass 'edit', '--non-interactive', $Name
```
The same kind of problem might occur on the edit operation than the below comment.
If the Edit fail, we would want to notify the user too.

TylerLeonhardt replied:
My last commit should fix this generically.

TylerLeonhardt committed:
better error handling

itfranck committed:
It should resolve all the problems forever ? (#16)

TylerLeonhardt commented:
![image](https://user-images.githubusercontent.com/2644648/99135361-b97ae480-25d6-11eb-98f9-5123a8f112e3.png)
:(

TylerLeonhardt commented:
@itfranck if we _ignore_ errors do we want to include them in the output stream or not?

TylerLeonhardt committed:
I think this is what was suppose to happen

TylerLeonhardt commented:
check my last commit, that makes all the Error's go away... however, the test still fails so something broke wrt PSCredentials.

itfranck commented:
> @itfranck if we _ignore_ errors do we want to include them in the output stream or not?

> check my last commit, that makes all the Error's go away... however, the test still fails so something broke wrt PSCredentials.
Oh yeah. Oops. That `-not` was very important 😊 


Does [that](https://github.com/TylerLeonhardt/SecretManagement.LastPass/blob/287772548d0c5dc778317c7319335dbe58caf2de/test/Extension.Tests.ps1#L60) work for you ? 

```
        $secret.Password | ConvertFrom-SecureString -AsPlainText | Should -BeExactly $secretText
```

I ran the test and had an error too there but switching that line for : 

```
[pscredential]::new('myUser', $secret.Password).GetNetworkCredential().Password
```
resolved the issue for me.
For me (Using PS 5.1 for module development), your statement gave me the gibberish secure string in plain text but still as a secure string (bunch of 0 and 1) 


edit: looks like I have another error there... still looking.

TylerLeonhardt commented:
Looks like the test failure was a fluke. It's working for me now!

itfranck reviewed (approved):
I think at this point this PR is solid and we got rid of all the initial issues.
